storage/innobase/handler/ha_innodb.cc

```diff
@@ -18799,10 +18755,45 @@ wsrep_innobase_kill_one_trx(
     DBUG_VOID_RETURN;
   }
 
-  /* Here we need to lock THD::LOCK_thd_data to protect from
-  concurrent usage or disconnect or delete. */
+  /* Grab reference to victim_trx before releasing the mutex, this will
+     prevent victim to release locks or commit while the mutex is
+     unlocked. The state may change to TRX_STATE_COMMITTED_IN_MEMORY.
+     See skip_lock_inheritance_n_ref in trx0trx.h. */
+  const trx_id_t victim_trx_id= victim_trx->id;
+retry_lock:
+  victim_trx->reference();
+  trx_mutex_exit(victim_trx);
```

Comment:
Why is `lock_sys.mutex` not being released during the retry loop? It should be a bigger bottleneck than `victim_trx->mutex`.

Reply:
Is it safe to release `lock_sys.mutex` here, this function is called from 
```
lock_rec_other_has_conflicting() -> wsrep_kill_victim() -> wsrep_innobase_kill_one_trx()
```
and
```
lock_table_other_has_incompatible() -> wsrep_kill_victim() -> wsrep_innobase_kill_one_trx()
```

Looking around `lock_rec_other_has_conflicting()` and `lock_table_other_has_incompatible()`, releasing `lock_sys.mutex` here does not look safe as the call happens in the middle of `lock_sys.rec_hash` iteration.